### ISSUE TITLE
panic: runtime: goroutine stack exceeds 1000000000-byte limit

### ISSUE THREAD

**Issue by markfra** [hashicorp/consul#4040]
### Description of the Issue (and unexpected/desired result)

We ran into an issue on 1 of our clusters this afternoon, which caused consul to panic.
When debugging the issue it appeared that someone by mistake set the hostname of 1 server to <hostname>.query.consul (the service for this hostname did not exist).

What followed was a panic on all (3 atm) consul servers and in the end a total outage of that cluster.

### Reproduction steps

server1 # enable query_templating with this tempate
```
 {
   "Name": "",
   "Template": {
     "Type": "name_prefix_match"
   },
   "Service": {
     "Service": "${name.full}"
   }
 }
```
```clien1t# hostname <something>.query.consul```

watch the cluster hang itself starting with the panic: runtime: goroutine stack exceeds 1000000000-byte limit.

### `consul version` for both Client and Server
Client: 0.7.2
Server: 0.7.5

### `consul info` for both Client and Server
Client:
```
agent:
	check_monitors = 3
	check_ttls = 0
	checks = 3
	services = 3
build:
	prerelease = 
	revision = 'a9afa0c
	version = 0.7.2
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 48
	max_procs = 1
	os = linux
	version = go1.7.3
serf_lan:
	encrypted = true
	event_queue = 0
	event_time = 246179
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 54272
	members = 2009
	query_queue = 0
	query_time = 4
```

Server:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 1
build:
	prerelease = 
	revision = '21f2d5a
	version = 0.7.5
consul:
	bootstrap = false
	known_datacenters = 4
	leader = false
	leader_addr = 10.98.72.142:8300
	server = true
raft:
	applied_index = 63014335
	commit_index = 63014335
	fsm_pending = 0
	last_contact = 50.020311ms
	last_log_index = 63014335
	last_log_term = 3856
	last_snapshot_index = 63008618
	last_snapshot_term = 3856
	latest_configuration = [{Suffrage:Voter ID:x.x.72.142:8300 Address:x.x.72.142:8300} {Suffrage:Voter ID:x.x.72.141:8300 Address:x.x.72.141:8300} {Suffrage:Voter ID:x.x.72.140:8300 Address:x.x.72.140:8300}]
	latest_configuration_index = 35892581
	num_peers = 2
	protocol_version = 1
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 3856
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 5199
	max_procs = 2
	os = linux
	version = go1.7.5
serf_lan:
	encrypted = true
	event_queue = 0
	event_time = 246179
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 54272
	members = 2009
	query_queue = 0
	query_time = 4
serf_wan:
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1245
	members = 12
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details
rhel7.4 for both client & server

### Log Fragments or Link to [gist](https://gist.github.com/)
https://gist.githubusercontent.com/markfra/a1deb7e75a18fc66fc755c3b204d9ca0/raw/0f7bff5a453f4e7ab870fbc8dd063b730fc42123/consul-panic-log

We assume the issue is caused when processing requests of the affected client, causing an infinite loop and panic in the end.
Obviously this was a user-error on our side and we will be taking measures to stop it from happening again, however the end-result is quite staggering with a whole cluster-outage.

We have a number of questions:

- We know our consul version is quite old, so is this perhaps solved in a newer version?
- Could this be solved at all on the consul-side, or is it 'working as intended'?
- Do you need more detail/debug-logging/etc?

**Comment by preetapan** [comment]
@markfra Thanks for reporting this. This is an edge case in our DNS resolver for prepared queries which end with the ".query.consul" domain by default, and we should handle it on the Consul side better.

**Comment by pierresouchay** [comment]
Might be fixed by https://github.com/hashicorp/consul/pull/4918

**Comment by kyhavlov** [comment]
I tried to reproduce this on master and couldn't - it seems to have been fixed by #4918. If anyone is still seeing this, please comment here and we can reopen this issue.
